src/java/org/apache/cassandra/db/marshal/VectorType.java

```diff
@@ -0,0 +1,603 @@
+        int separator = comparableBytes.next();
+        while (separator != ByteSource.TERMINATOR)
+        {
+            buffers.add(elementType.fromComparableBytes(accessor, comparableBytes, version));
+            separator = comparableBytes.next();
+        }
+        return decomposeRaw(buffers, accessor);
+    }
+
+    @Override
+    public CQL3Type asCQL3Type()
+    {
+        return new CQL3Type.Vector(this);
+    }
+
+    public AbstractType<T> getElementsType()
+    {
+        return elementType;
+    }
+
+    // vector of nested types is hard to parse, so fall back to bytes string matching ListType
+    @Override
+    public <V> String getString(V value, ValueAccessor<V> accessor)
+    {
+        return BytesType.instance.getString(value, accessor);
+    }
+
+    @Override
+    public ByteBuffer fromString(String source) throws MarshalException
+    {
+        try
+        {
+            return ByteBufferUtil.hexToBytes(source);
+        }
+        catch (NumberFormatException e)
+        {
+            throw new MarshalException(String.format("cannot parse '%s' as hex bytes", source), e);
+        }
+    }
+
+    @Override
+    public List<AbstractType<?>> subTypes()
+    {
+        return Collections.singletonList(elementType);
+    }
+
+    @Override
+    public String toJSONString(ByteBuffer buffer, ProtocolVersion protocolVersion)
+    {
+        return toJSONString(buffer, ByteBufferAccessor.instance, protocolVersion);
+    }
+
+    @Override
+    public <V> String toJSONString(V value, ValueAccessor<V> accessor, ProtocolVersion protocolVersion)
+    {
+        StringBuilder sb = new StringBuilder();
+        sb.append('[');
+        List<V> split = split(value, accessor);
+        for (int i = 0; i < dimension; i++)
+        {
+            if (i > 0)
+                sb.append(", ");
+            sb.append(elementType.toJSONString(split.get(i), accessor, protocolVersion));
+        }
+        sb.append(']');
+        return sb.toString();
+    }
+
+    @Override
+    public Term fromJSONObject(Object parsed) throws MarshalException
+    {
+        if (parsed instanceof String)
+            parsed = JsonUtils.decodeJson((String) parsed);
+
+        if (!(parsed instanceof List))
+            throw new MarshalException(String.format(
+            "Expected a list, but got a %s: %s", parsed.getClass().getSimpleName(), parsed));
+
+        List<?> list = (List<?>) parsed;
+        if (list.size() != dimension)
+            throw new MarshalException(String.format("List had incorrect size: expected %d but given %d; %s", dimension, list.size(), list));
+        List<Term> terms = new ArrayList<>(list.size());
+        for (Object element : list)
+        {
+            if (element == null)
+                throw new MarshalException("Invalid null element in list");
+            terms.add(elementType.fromJSONObject(element));
+        }
+
+        return new Vectors.DelayedValue(this, terms);
```

Comment:
nit: `Vectors.DelayedValue` is parameterized so this should be `Vectors.DelayedValue<>`. Apart from anything this removes another warning.